src/transaction_sender/service.zig

```diff
@@ -0,0 +1,349 @@
+            }
+        }
+
+        if (transaction_batch.count() >= config.batch_size or
+            (transaction_batch.count() > 0 and
+            (try last_batch_sent.elapsed()).asNanos() >= config.batch_send_rate.asNanos()))
+        {
+            try sendTransactions(
+                allocator,
+                sender,
+                config,
+                cluster_info_rw,
+                transaction_batch.values(),
+            );
+            last_batch_sent = try Instant.now();
+
+            var transaction_pool: *TransactionPool, var transaction_pool_lock = transaction_pool_rw.writeWithLock();
+            defer transaction_pool_lock.unlock();
+
+            logger.infof("Adding {d} new transactions to pool.", .{transaction_batch.count()});
+            for (transaction_batch.values()) |_tx| {
+                if (transaction_pool.count() >= config.pool_max_size) {
+                    logger.warnf("Transaction pool is full, dropping transaction: signature={s}", .{_tx.signature});
+                    continue;
+                }
+                var tx = _tx; // Is there a nicer way to do this?
+                tx.last_sent_time = last_batch_sent;
+                try transaction_pool.put(tx.signature, tx);
+            }
+            transaction_batch.clearRetainingCapacity();
+        }
+    }
+}
+
+fn processTransactionsThread(
+    allocator: std.mem.Allocator,
+    sender: *Channel(std.ArrayList(Packet)),
+    config: Config,
+    cluster_info_rw: *RwMux(LeaderInfo),
+    transaction_pool_rw: *RwMux(TransactionPool),
+    exit: *AtomicBool,
+    logger: Logger,
+) !void {
+    errdefer exit.store(true, .unordered);
+
+    while (!exit.load(.unordered)) {
+        std.time.sleep(config.pool_process_rate.asNanos());
+
+        var transaction_pool: *TransactionPool, var transaction_pool_lock = transaction_pool_rw.writeWithLock();
+        defer transaction_pool_lock.unlock();
+
+        if (transaction_pool.count() == 0) continue;
+
+        try processTransactions(
+            allocator,
+            sender,
+            config,
+            cluster_info_rw,
+            transaction_pool,
+            logger,
+        );
+    }
+}
+
+fn processTransactions(
+    allocator: std.mem.Allocator,
+    sender: *Channel(std.ArrayList(Packet)),
+    config: Config,
+    cluster_info_rw: *RwMux(LeaderInfo),
+    transaction_pool: *TransactionPool,
+    logger: Logger,
+) !void {
+    var successful_signatures = std.ArrayList(Signature).init(allocator);
+    defer successful_signatures.deinit();
+
+    var retry_signatures = std.ArrayList(Signature).init(allocator);
+    defer retry_signatures.deinit();
+
+    var drop_signatures = std.ArrayList(Signature).init(allocator);
+    defer drop_signatures.deinit();
+
+    var rpc_arena = std.heap.ArenaAllocator.init(allocator);
+    defer rpc_arena.deinit();
+
+    var rpc_client = RpcClient.init(allocator, .Testnet);
+    defer rpc_client.deinit();
```

Comment:
probs want to re-use the memory allocated on each of these by pulling them out of the loop and use clearRetainingCapacity